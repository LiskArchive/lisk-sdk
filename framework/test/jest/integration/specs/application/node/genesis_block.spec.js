/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { getAddressFromPublicKey } = require('@liskhq/lisk-cryptography');
const {
	chainUtils,
	storageUtils,
	configUtils,
} = require('../../../../../utils');
const delegateListForTheFirstRound = require('../../../../../fixtures/config/devnet/delegates_for_first_round.json');
const genesisBlock = require('../../../../../fixtures/config/devnet/genesis_block');

describe('genesis block', () => {
	const dbName = 'genesis_block';
	const TRANSACTION_TYPE_DELEGATE_REGISTRATION = 10;
	let storage;
	let chainModule;

	beforeAll(async () => {
		storage = new storageUtils.StorageSandbox(
			configUtils.storageConfig({ database: dbName }),
			dbName,
		);
		await storage.bootstrap();
		chainModule = await chainUtils.createAndLoadChainModule(dbName);
	});

	afterAll(async () => {
		await chainModule.unload();
		await storage.cleanup();
	});

	describe('given the application has not been initialized', () => {
		describe('when chain module is bootstrapped', () => {
			it('should save genesis block to the database', async () => {
				const block = await storageUtils.getBlock(storage, genesisBlock.id);
				expect(block.id).toEqual(genesisBlock.id);
				expect(block.height).toEqual(1);
			});

			it('should have genesis transactions in database', async () => {
				const block = await storageUtils.getBlock(storage, genesisBlock.id);
				const ids = genesisBlock.transactions.map(t => t.id);
				const allExist = ids.every(id =>
					block.transactions.map(tx => tx.id).includes(id),
				);

				expect(allExist).toEqual(true);
			});

			it('should save accounts for the registered genesis delegates', async () => {
				// Get accounts of delegate registeration
				const delegateRegistrationTransactions = genesisBlock.transactions.filter(
					transaction =>
						transaction.type === TRANSACTION_TYPE_DELEGATE_REGISTRATION,
				);
				const delegateAccountsAddressesInGenesisBlock = delegateRegistrationTransactions.map(
					transaction => getAddressFromPublicKey(transaction.senderPublicKey),
				);
				// Get delegate accounts in genesis block from the database
				const accountsFromDb = await Promise.all(
					delegateAccountsAddressesInGenesisBlock.map(address =>
						storageUtils.getAccount(storage, address),
					),
				);
				const allAccountsAreDelegate = delegateAccountsAddressesInGenesisBlock.every(
					address =>
						accountsFromDb.find(account => address === account.address),
				);

				expect(allAccountsAreDelegate).toEqual(true);
			});

			it('should have correct vote weight for genesis delegates', async () => {
				// All delegates has vote weight of total supply
				const voteWeightOfDevnetDelegates = '10000000000000000';
				// Get accounts of delegate registeration
				const delegateRegistrationTransactions = genesisBlock.transactions.filter(
					transaction =>
						transaction.type === TRANSACTION_TYPE_DELEGATE_REGISTRATION,
				);
				const delegateAccountsAddressesInGenesisBlock = delegateRegistrationTransactions.map(
					transaction => getAddressFromPublicKey(transaction.senderPublicKey),
				);
				// Get delegate accounts in genesis block from the database
				const accountsFromDb = await Promise.all(
					delegateAccountsAddressesInGenesisBlock.map(address =>
						storageUtils.getAccount(storage, address),
					),
				);
				const allAccountsHaveCorrectVoteWeight = delegateAccountsAddressesInGenesisBlock.every(
					address =>
						accountsFromDb.find(
							account =>
								address === account.address &&
								account.voteWeight === voteWeightOfDevnetDelegates,
						),
				);

				expect(allAccountsHaveCorrectVoteWeight).toEqual(true);
			});

			it('should have correct delegate list', async () => {
				const delegateListFromChain = await chainUtils.getDelegateList(
					chainModule.chain,
					1,
				);
				expect(delegateListFromChain).toEqual(delegateListForTheFirstRound);
			});
		});
	});

	describe('given the application has been initialized previously', () => {
		describe('when chain module is bootstrapped', () => {
			it('should have genesis transactions in database', async () => {
				const block = await storageUtils.getBlock(storage, genesisBlock.id);
				const ids = genesisBlock.transactions.map(t => t.id);
				const allExist = ids.every(id =>
					block.transactions.map(tx => tx.id).includes(id),
				);

				expect(allExist).toEqual(true);
			});

			it('should save accounts for the registered genesis delegates', async () => {
				// Get accounts of delegate registeration
				const delegateRegistrationTransactions = genesisBlock.transactions.filter(
					transaction =>
						transaction.type === TRANSACTION_TYPE_DELEGATE_REGISTRATION,
				);
				const delegateAccountsAddressesInGenesisBlock = delegateRegistrationTransactions.map(
					transaction => getAddressFromPublicKey(transaction.senderPublicKey),
				);
				// Get delegate accounts in genesis block from the database
				const accountsFromDb = await Promise.all(
					delegateAccountsAddressesInGenesisBlock.map(address =>
						storageUtils.getAccount(storage, address),
					),
				);
				const allAccountsAreDelegate = delegateAccountsAddressesInGenesisBlock.every(
					address =>
						accountsFromDb.find(account => address === account.address),
				);

				expect(allAccountsAreDelegate).toEqual(true);
			});

			it('should have correct vote weight for genesis delegates', async () => {
				// All delegates has vote weight of total supply
				const voteWeightOfDevnetDelegates = '10000000000000000';
				// Get accounts of delegate registeration
				const delegateRegistrationTransactions = genesisBlock.transactions.filter(
					transaction =>
						transaction.type === TRANSACTION_TYPE_DELEGATE_REGISTRATION,
				);
				const delegateAccountsAddressesInGenesisBlock = delegateRegistrationTransactions.map(
					transaction => getAddressFromPublicKey(transaction.senderPublicKey),
				);
				// Get delegate accounts in genesis block from the database
				const accountsFromDb = await Promise.all(
					delegateAccountsAddressesInGenesisBlock.map(address =>
						storageUtils.getAccount(storage, address),
					),
				);
				const allAccountsHaveCorrectVoteWeight = delegateAccountsAddressesInGenesisBlock.every(
					address =>
						accountsFromDb.find(
							account =>
								address === account.address &&
								account.voteWeight === voteWeightOfDevnetDelegates,
						),
				);

				expect(allAccountsHaveCorrectVoteWeight).toEqual(true);
			});

			it('should have correct delegate list', async () => {
				const delegateListFromChain = await chainUtils.getDelegateList(
					chainModule.chain,
					1,
				);
				expect(delegateListFromChain).toEqual(delegateListForTheFirstRound);
			});
		});
	});
});
