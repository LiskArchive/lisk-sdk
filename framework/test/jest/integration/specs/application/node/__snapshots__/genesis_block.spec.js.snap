// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genesis block given the application has been initialized previously when chain module is bootstrapped should have correct delegate list 1`] = `
Array [
  "13782190884886479261L",
  "1664315804091957933L",
  "1062642951156140585L",
  "16308812918716033341L",
  "5306771970976769752L",
  "3165313486832479357L",
  "15669876130162831140L",
  "8592552312055292576L",
  "2511141319964500017L",
  "12905869062797428148L",
  "14560629400712917194L",
  "2758933299398072267L",
  "15657488085702919962L",
  "18070013346623491378L",
  "4706931147950389586L",
  "8902618697721189524L",
  "14698332900351700944L",
  "7700165370820050502L",
  "11326636313532218837L",
  "13926502931445220391L",
  "7260717755366070561L",
  "457266766135430879L",
  "17726113224450398596L",
  "3868369556376898434L",
  "10045031187186962062L",
  "18131785163125373329L",
  "10452881617068866990L",
  "12424016829136501093L",
  "14235724695859131520L",
  "10016685355739180605L",
  "5304144955874014073L",
  "6317532335899444681L",
  "17068238285058560856L",
  "7767444336222298443L",
  "10431315846496304288L",
  "6177803802995165744L",
  "5261458975214470789L",
  "8762414756737057545L",
  "2252478578888840763L",
  "8158286759636279504L",
  "17749755823863372060L",
  "7910234169285432058L",
  "923992554593700306L",
  "13824100553290591171L",
  "3479602363991858353L",
  "11750255083444888021L",
  "15360529257123074396L",
  "13650664016494124634L",
  "5670277688518281662L",
  "15686825971214030955L",
  "11189372663423096911L",
  "13803933794686825569L",
  "9164804013838025941L",
  "14029548277628731563L",
  "12727144553591870062L",
  "12616371792697673866L",
  "3426690280983981237L",
  "7668353963380076435L",
  "11053440961949385483L",
  "2375380535924104331L",
  "592887264274404011L",
  "17787178302334615844L",
  "17088641110630230851L",
  "3885152915127841161L",
  "16936666638951007157L",
  "1696044429496474122L",
  "774271798011194039L",
  "16900039088698463203L",
  "4494601661524203864L",
  "2092334120572876076L",
  "18363567119458536866L",
  "2337346334401707579L",
  "3854561502400413233L",
  "13203042986869846881L",
  "14368852892347477005L",
  "15384849944947270279L",
  "17863689263870159728L",
  "18280785839802205865L",
  "6773911674068309958L",
  "7786289702094844687L",
  "9077548379631877989L",
  "9447508130077835324L",
  "6091819248619420783L",
  "61015361474842367L",
  "235626005720150263L",
  "14119664716639126171L",
  "16269561247929808361L",
  "15807287666952782927L",
  "8531579280410192796L",
  "14299179756080395942L",
  "10882030944401356291L",
  "15621640699942494333L",
  "8768586841172786527L",
  "1810963503742170588L",
  "11958215799471943779L",
  "2239791898636671159L",
  "8037225175530047310L",
  "8167345791948290988L",
  "6445240216714680253L",
  "8236718713333257698L",
  "6050619090308352211L",
  "13689625589582910799L",
  "17718487952107855818L",
]
`;

exports[`genesis block given the application has not been initialized when chain module is bootstrapped should have correct delegate list 1`] = `
Array [
  "13782190884886479261L",
  "1664315804091957933L",
  "1062642951156140585L",
  "16308812918716033341L",
  "5306771970976769752L",
  "3165313486832479357L",
  "15669876130162831140L",
  "8592552312055292576L",
  "2511141319964500017L",
  "12905869062797428148L",
  "14560629400712917194L",
  "2758933299398072267L",
  "15657488085702919962L",
  "18070013346623491378L",
  "4706931147950389586L",
  "8902618697721189524L",
  "14698332900351700944L",
  "7700165370820050502L",
  "11326636313532218837L",
  "13926502931445220391L",
  "7260717755366070561L",
  "457266766135430879L",
  "17726113224450398596L",
  "3868369556376898434L",
  "10045031187186962062L",
  "18131785163125373329L",
  "10452881617068866990L",
  "12424016829136501093L",
  "14235724695859131520L",
  "10016685355739180605L",
  "5304144955874014073L",
  "6317532335899444681L",
  "17068238285058560856L",
  "7767444336222298443L",
  "10431315846496304288L",
  "6177803802995165744L",
  "5261458975214470789L",
  "8762414756737057545L",
  "2252478578888840763L",
  "8158286759636279504L",
  "17749755823863372060L",
  "7910234169285432058L",
  "923992554593700306L",
  "13824100553290591171L",
  "3479602363991858353L",
  "11750255083444888021L",
  "15360529257123074396L",
  "13650664016494124634L",
  "5670277688518281662L",
  "15686825971214030955L",
  "11189372663423096911L",
  "13803933794686825569L",
  "9164804013838025941L",
  "14029548277628731563L",
  "12727144553591870062L",
  "12616371792697673866L",
  "3426690280983981237L",
  "7668353963380076435L",
  "11053440961949385483L",
  "2375380535924104331L",
  "592887264274404011L",
  "17787178302334615844L",
  "17088641110630230851L",
  "3885152915127841161L",
  "16936666638951007157L",
  "1696044429496474122L",
  "774271798011194039L",
  "16900039088698463203L",
  "4494601661524203864L",
  "2092334120572876076L",
  "18363567119458536866L",
  "2337346334401707579L",
  "3854561502400413233L",
  "13203042986869846881L",
  "14368852892347477005L",
  "15384849944947270279L",
  "17863689263870159728L",
  "18280785839802205865L",
  "6773911674068309958L",
  "7786289702094844687L",
  "9077548379631877989L",
  "9447508130077835324L",
  "6091819248619420783L",
  "61015361474842367L",
  "235626005720150263L",
  "14119664716639126171L",
  "16269561247929808361L",
  "15807287666952782927L",
  "8531579280410192796L",
  "14299179756080395942L",
  "10882030944401356291L",
  "15621640699942494333L",
  "8768586841172786527L",
  "1810963503742170588L",
  "11958215799471943779L",
  "2239791898636671159L",
  "8037225175530047310L",
  "8167345791948290988L",
  "6445240216714680253L",
  "8236718713333257698L",
  "6050619090308352211L",
  "13689625589582910799L",
  "17718487952107855818L",
]
`;
